mleap-runtime/src/main/scala/ml/combust/mleap/runtime/LeapFrame.scala

```diff
@@ -50,63 +52,69 @@ trait LeapFrame[LF <: LeapFrame[LF]] extends TransformBuilder[LF] with Serializa
+      schema.indexedField(name).flatMap {
+        case (index, field) =>
+          if(dataType.fits(field.dataType)) {
+            Try(r => r.get(index))
+          } else {
+            Failure(new IllegalArgumentException(s"field $name data type ${field.dataType} does not match $dataType"))
+          }
+      }
+    case ArraySelector(fields @ _*) =>
+      (for(base <- Try(dataType.asInstanceOf[ListType].base)) yield {
+        var i = 0
+        fields.foldLeft(Try(new Array[Int](fields.length))) {
+          case (indices, name) =>
+            schema.indexedField(name).flatMap {
+              case (index, field) =>
+                if (base.fits(field.dataType)) {
+                  indices.map {
+                    ids =>
+                      ids(i) = index
+                      i = i + 1
+                      ids
+                  }
+                } else {
+                  Failure(new IllegalArgumentException(s"field $name data type ${field.dataType} does not match $base"))
+                }
+            }
+        }.map(indices => (r: Row) => indices.map(i => r.get(i)))
+      }).flatMap(identity)
+  }
 
   /** Try to add a field to the LeapFrame.
     *
     * Returns a Failure if trying to add a field that already exists.
     *
-    * @param field field to add
-    * @param f function for calculating field value
+    * @param name name of field
+    * @param selectors row selectors used to generate inputs to udf
+    * @param udf user defined function for calculating field value
     * @return try new LeapFrame with new field
     */
-  def withField(field: StructField)
-               (f: (Row) => Any): Try[LF] = {
-    schema.withField(field).map {
-      schema2 =>
-        val dataset2 = dataset.withValue(f)
-        withSchemaAndDataset(schema2, dataset2)
-    }
-  }
-
-  /** Try to add multiple fields to the LeapFrame.
-    *
-    * Returns a Failure if trying to add any existing fields.
-    *
-    * @param field first field to add
-    * @param fields fields to add
-    * @param f function for calculating new field values
-    * @return try new LeapFrame with new fields
-    */
-  def withFields(field: StructField, fields: StructField *)
-                (f: (Row) => Row): Try[LF] = {
-    withFields(field +: fields)(f)
-  }
-
-  /** Try to add multiple fields to the LeapFrame.
-    *
-    * Returns a Failure if trying to add any existing fields.
-    *
-    * @param fields fields to add
-    * @param f function for calculating new field values
-    * @return try new LeapFrame with new fields
-    */
-  def withFields(fields: Seq[StructField])
-                (f: (Row) => Row): Try[LF] = {
-    schema.withFields(fields).map {
-      schema2 =>
-        val dataset2 = dataset.withValues(f)
-        withSchemaAndDataset(schema2, dataset2)
+  def withField(name: String, selectors: Selector *)
+               (udf: UserDefinedFunction): Try[LF] = {
+    var i = 0
+    selectors.foldLeft(Try(Seq[RowSelector]())) {
```

Comment:
This will cause allocations of unneeded tuples, which I want to avoid when possible. Otherwise, I am all for completely functional programming :)